ostd/src/device/dispatcher/io_mem.rs

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! I/O Memory access dispatcher.
+
+use alloc::vec::Vec;
+use core::ops::Range;
+
+use log::debug;
+use spin::Once;
+
+use crate::{
+    device::io_mem::IoMem,
+    mm::{CachePolicy, PageFlags},
+    util::vaddr_alloc::VirtAddrAllocator,
+};
+
+/// I/O memory dispatcher that dispatches memory I/O access to device drivers.
+pub struct IoMemDispatcher {
+    allocators: Vec<VirtAddrAllocator>,
+}
+
+impl IoMemDispatcher {
+    /// Gets the I/O memory access for `range`.
+    ///
+    /// If the range is not available, then the return value will be `None`.
+    pub fn get(&self, range: Range<usize>) -> Option<IoMem> {
+        find_allocator(&self.allocators, &range)?
+            .alloc_specific(&range)
+            .ok()?;
+
+        debug!("Getting MMIO range:{:x?}..{:x?}", range.start, range.end);
+
+        // SAFETY: The created `IoMem` is guaranteed not to access physical memory or system device I/O.
+        unsafe { Some(IoMem::new(range, PageFlags::RW, CachePolicy::Uncacheable)) }
+    }
+
+    /// Recycles an MMIO range.
+    #[allow(dead_code)]
+    pub(crate) fn recycle(&self, range: Range<usize>) {
```

Comment:
1. The `recycle` has to be `unsafe` because it requires the caller to have ownership of the MMIO region. (Otherwise it will break the type invariant, i.e. the MMIO region of the system device cannot be put into the dispatcher).
 2. Given the above, we should just unwrap/panic if the `find_allocator` fails during `recycle()` (if the caller respects the safety requirements, it should _never_ fail).
 3. https://github.com/asterinas/asterinas/pull/1479#discussion_r1827779785 isn't really addressed, I think. (Although it's just a minor style comment, and it's fine to ignore it if you don't think it makes the code better).